### PR TITLE
feat: add resize onnx to tosa

### PR DESCRIPTION
**Pull request by maxbartel**
This introduces the conversion of the `ResizeOp `from ONNX to TOSA. 
The different modes of the op would all probably deserve their own operation instead of just an `Attribute`, so be extra careful when reviewing please. Also TOSA has the concept of using a fraction instead a float for the scale. You can look it up here: https://www.mlplatform.org/tosa/tosa_spec.html#_resize. I tried to give as much information as possible on how I arrived at the different formulas. Please let me know if something is not clear.

We also have to disallow a lot of cases because the onnx op is more powerful than the TOSA one.

### CODE REVIEW

File: src/Conversion/ONNXToTOSA/Tensor/Resize.cpp [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+#include <cstdint>
+#include <numeric>
+
+using namespace mlir;
+
+namespace onnx_mlir {
+
+namespace {
+
+std::tuple<int64_t, int64_t> floatToFraction(float number) {
```
Comment (ttjost) [inline src/Conversion/ONNXToTOSA/Tensor/Resize.cpp:28]: Tuple are smelly (https://arne-mertz.de/2017/03/smelly-pair-tuple/), maybe you can make a `struct`? Something like:
```suggestion
struct fractionNumber{
int64_t numerator;
int64_t denominator;
};
```

File: src/Conversion/ONNXToTOSA/Tensor/Resize.cpp [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+
+    // Get axis values if set. Default is all axis in normal order.
+    llvm::SmallVector<int64_t> axisVec;
+    if (axis.has_value()) {
+      valuesFromAxis(&axis.value(), axisVec);
+    } else {
+      axisVec.append({0, 1, 2, 3});
+    }
+
+    // Set these explicitly just out of convinience.
```
Comment (ttjost) [inline src/Conversion/ONNXToTOSA/Tensor/Resize.cpp:167]:
```suggestion
    // Set these explicitly just out of convenience.
```

File: src/Conversion/ONNXToTOSA/Tensor/Resize.cpp [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+      // Even if the shapes are identical, a scale value other than 1 is
+      // possible. TOSA does not allow that for non-spatial dimensions.
+      for (int64_t nonSpatialDimensions : {0, 1}) {
+        if (scales[nonSpatialDimensions] != 1)
+          return rewriter.notifyMatchFailure(
+              op, "Axis Attr if present must contain both output dimesions.");
+      }
+
+      // In TOSA the scale is a fraction of two integer numbers.
+      std::tie(outputHeight, inputHeight) = floatToFraction(scales[2]);
```
Comment (ttjost) [inline src/Conversion/ONNXToTOSA/Tensor/Resize.cpp:198]: `auto` maybe?
```suggestion
      auto [outputHeight, inputHeight] = floatToFraction(scales[2]);
```

**Review by ttjost — COMMENTED** [review]
I still want to go through it again, but that should be a good start. Thanks!

File: src/Conversion/ONNXToTOSA/Tensor/Resize.cpp [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+
+    if (halfPixelSymmetric)
+      return rewriter.notifyMatchFailure(op,
+          "TOSA does not support float offsets which are required "
+          "for symmetric mode.");
+
+    // Adapted from TFL to TOSA.
+    auto normalize = [&](int64_t output, int64_t input) {
+      // Dimension is length 1, we are just sampling from one value.
+      int64_t n, d, offset, border;
```
Comment (flemairen6) [inline src/Conversion/ONNXToTOSA/Tensor/Resize.cpp:221]: Could we have better naming than n and d there, makes the code quite harder to read when you don't know what they are about

File: src/Conversion/ONNXToTOSA/Tensor/Resize.cpp (hunk 3 above)
Comment (maxbartel) [reply src/Conversion/ONNXToTOSA/Tensor/Resize.cpp:198]: Changes this to a struct, too

File: src/Conversion/ONNXToTOSA/Tensor/Resize.cpp [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+      if (isNearest && isNearestModeFloor) {
+        offset -= n / 2;
+      }
+
+      // We can compute this directly based on previous values.
+      border = d * (output - 1) - n * (input - 1) + offset;
+      return std::make_tuple(n, d, offset, border);
+    };
+
+    auto [scale_y_n, scale_y_d, offset_y, border_y] =
```
[the remark this answers is not in the document]
Comment (maxbartel) [reply src/Conversion/ONNXToTOSA/Tensor/Resize.cpp:255]: This was there because of the copy from TF. They use the Google Coding style. I changed this together with the struct

File: src/Conversion/ONNXToTOSA/Tensor/Resize.cpp [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+// =============================================================================
+//
+// This file lowers the ONNX Resize operator to TOSA dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"
+#include "src/Conversion/ONNXToTOSA/DialectBuilder.hpp"
+#include "src/Conversion/ONNXToTOSA/ONNXToTOSACommon.hpp"
+#include "src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp"
```
Comment (cmcgirr-amd) [inline src/Conversion/ONNXToTOSA/Tensor/Resize.cpp:18]: nit: header grouping https://llvm.org/docs/CodingStandards.html#include-style

File: src/Conversion/ONNXToTOSA/Tensor/Resize.cpp [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+    // With only static dimensions, scales and sizes as inputs are not relevant
+    // anymore.
+    if (inputType.isDynamicDim(2) || inputType.isDynamicDim(3)) {
+      return rewriter.notifyMatchFailure(
+          resizeOp, "Only static sized tensors are supported.");
+    }
+
+    if (mode == "cubic") {
+      return rewriter.notifyMatchFailure(
+          resizeOp, "TOSA does not support cubic interpolation.");
```
Comment (cmcgirr-amd) [inline src/Conversion/ONNXToTOSA/Tensor/Resize.cpp:148]: can you add a case for this in the test?

File: src/Conversion/ONNXToTOSA/Tensor/Resize.cpp [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+    auto outputShape = resultType.getShape();
+
+    if (inputShape[0] != outputShape[0] || inputShape[1] != outputShape[1])
+      return rewriter.notifyMatchFailure(
+          op, "Cannot resize non spatial dimensions.");
+
+    // Get axis values if set. Default is all axis in normal order.
+    llvm::SmallVector<int64_t> axisVec;
+    if (axis.has_value()) {
+      valuesFromAxis(&axis.value(), axisVec);
```
Comment (cmcgirr-amd) [inline src/Conversion/ONNXToTOSA/Tensor/Resize.cpp:177]: I wonder how strict we should be here because behavior is undefined if an axis is repeated and the ONNX Resize verifier does not check for that particular case. What do you think?

File: src/Conversion/ONNXToTOSA/Tensor/Resize.cpp [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+      for (auto [index, value] : llvm::enumerate(axisVec))
+        scales[value] =
+            elementsAttr.getValues<FloatAttr>()[index].getValueAsDouble();
+
+      // Even if the shapes are identical, a scale value other than 1 is
+      // possible. TOSA does not allow that for non-spatial dimensions.
+      for (int64_t nonSpatialDimensions : {0, 1}) {
+        if (scales[nonSpatialDimensions] != 1)
+          return rewriter.notifyMatchFailure(
+              op, "Axis Attr if present must contain both output dimesions.");
```
Comment (cmcgirr-amd) [inline src/Conversion/ONNXToTOSA/Tensor/Resize.cpp:209]:
```suggestion
              op, "Axis Attr if present must contain both output dimensions.");
```
typo

File: src/Conversion/ONNXToTOSA/Tensor/Resize.cpp [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+  struct ScaleHelper {
+    ScaleHelper(
+        int64_t numerator, int64_t denominator, int64_t offset, int64_t border)
+        : numerator(numerator), denominator(denominator), offset(offset),
+          border(border){};
+    int64_t numerator, denominator, offset, border;
+  };
+
+  struct FractionNumber {
+    FractionNumber(int64_t numerator, int64_t denominator)
```
Comment (cmcgirr-amd) [inline src/Conversion/ONNXToTOSA/Tensor/Resize.cpp:50]: super nit: do we need this constructor if it is unused?

File: src/Conversion/ONNXToTOSA/Tensor/Resize.cpp [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+    StringRef resizeMode = isBilinear ? "BILINEAR" : "NEAREST_NEIGHBOR";
+
+    if (halfPixelSymmetric)
+      return rewriter.notifyMatchFailure(op,
+          "TOSA does not support float offsets which are required "
+          "for symmetric mode.");
+
+    // Adapted from TFL to TOSA.
+    auto normalize = [&](int64_t output, int64_t input) {
+      // Dimension is length 1, we are just sampling from one value.
```
Comment (cmcgirr-amd) [inline src/Conversion/ONNXToTOSA/Tensor/Resize.cpp:239]: nit: maybe the comment should be moved down to line 242

File: src/Conversion/ONNXToTOSA/Tensor/Resize.cpp [hunk 12]
```diff
@@ -0,0 +1,314 @@
...
+    auto resizeOp = llvm::cast<ONNXResizeOp>(op);
+    Location loc = op->getLoc();
+    OpAdaptor adaptor(operands, op->getAttrDictionary());
+
+    TosaBuilder tosaBuilder(rewriter, loc);
+
+    Value input = adaptor.getX();
+    auto inputType = input.getType().dyn_cast<RankedTensorType>();
+
+    auto resultType =
```
Comment (cmcgirr-amd) [inline src/Conversion/ONNXToTOSA/Tensor/Resize.cpp:124]: I think we should also check that ONNX is using the supported types in the TOSA specification. `in_t` only supports `i8, i16, fp16, bf16 and fp32` and it seems to depend on the mode. 

https://www.mlplatform.org/tosa/tosa_spec.html#_resize

Comment (cmcgirr-amd) [reply src/Conversion/ONNXToTOSA/Tensor/Resize.cpp:124]: because onnx also supports unsigned in resize.

File: src/Conversion/ONNXToTOSA/Tensor/Resize.cpp [hunk 13]
```diff
@@ -0,0 +1,314 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===---------------- Resize.cpp - Resize Op-------------------------------===//
+//
+// Copyright (c) 2022 Advanced Micro Devices, Inc.
```
Comment (cmcgirr-amd) [inline src/Conversion/ONNXToTOSA/Tensor/Resize.cpp:7]:
```suggestion
// Copyright (c) 2023 Advanced Micro Devices, Inc.
```

File: src/Conversion/ONNXToTOSA/Tensor/Resize.cpp [hunk 14]
```diff
@@ -0,0 +1,314 @@
...
+  /// - scale_x_n = output_size
+  /// - offset = 0
+  ///
+  /// ## nearest_mode ##
+  /// If mode == nearest, then ONNX can set the nearest_mode attribute.
+  /// The standard case for TOSA of round_half_up. Support for floor can
+  /// be achieved with:
+  /// offset_x -= n/2
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
```
Comment (cmcgirr-amd) [inline src/Conversion/ONNXToTOSA/Tensor/Resize.cpp:113]: Another suggestion, could we break up this matchAndRewrite into small parts. I find it hard to follow at times. I think it may help to break it up into pieces, i.e. smaller functions. Because at the moment its a 200 line function. 

What do you think?

**Review by cmcgirr-amd — APPROVED** [review]

File: src/Conversion/ONNXToTOSA/Tensor/Resize.cpp (hunk 8 above)
Comment (maxbartel) [reply src/Conversion/ONNXToTOSA/Tensor/Resize.cpp:177]: Hmm I would say this kind of checking should happen in the verifier and not in the conversion. The verifiers in onnx-mlir are not super mature, maybe we should invest in them if we see a failure that could have been noticed earlier

File: src/Conversion/ONNXToTOSA/Tensor/Resize.cpp (hunk 14 above)
Comment (maxbartel) [reply src/Conversion/ONNXToTOSA/Tensor/Resize.cpp:113]: I will move the lambda and the scale value handling into the own functions. The lambda will get a lot more arguments, but I guess that is still better than now

File: src/Conversion/ONNXToTOSA/Tensor/Resize.cpp (hunk 11 above)
Comment (maxbartel) [reply src/Conversion/ONNXToTOSA/Tensor/Resize.cpp:239]: Ah that was for a deleted case, I will remove the comment altogether.

File: src/Conversion/ONNXToTOSA/Tensor/Resize.cpp (hunk 12 above)
Comment (maxbartel) [reply src/Conversion/ONNXToTOSA/Tensor/Resize.cpp:124]: Yeah so basically TOSA allows everything...
